src/index/TextScoring.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2024, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Felix Meisen (fesemeisen@outlook.de)
+
+#include "index/TextScoring.h"
+
+#include "index/Index.h"
+
+// ____________________________________________________________________________
+void ScoreData::calculateScoreData(const string& docsFileName,
+                                   bool addWordsFromLiterals,
+                                   const Index::TextVocab& textVocab,
+                                   const Index::Vocab& vocab) {
+  // Skip calculation if scoring mode is set to count
+  if (scoringMetric_ == TextScoringMetric::COUNT) {
+    return;
+  }
+
+  invertedIndex_.reserve(textVocab.size());
+  DocumentIndex docId = DocumentIndex::make(0);
+  DocsFileParser docsFileParser(docsFileName, textVocab.getLocaleManager());
+
+  // Parse the docsfile first if it exists
+  for (auto line : docsFileParser) {
+    // Set docId
+    docId = line.docId_;
+    // Parse docText for invertedIndex
+    addDocumentOrLiteralToScoreDataInvertedIndex(line.docContent_, docId,
+                                                 textVocab);
+  }
+
+  // Parse literals if added
+  if (!addWordsFromLiterals) {
+    return;
+  }
+  for (VocabIndex index = VocabIndex::make(0); index.get() < vocab.size();
+       index = index.incremented()) {
+    auto text = vocab[index];
+    if (!vocab.isLiteral(index)) {
+      continue;
+    }
+    // Reset parameters for loop
+    docId = docId.incremented();
+    std::string_view textView = text;
+
+    // Parse words in literal
+    addDocumentOrLiteralToScoreDataInvertedIndex(textView, docId, textVocab);
+  }
+}
+
+// ____________________________________________________________________________
+void ScoreData::addDocumentOrLiteralToScoreDataInvertedIndex(
+    std::string_view text, DocumentIndex docId,
+    const Index::TextVocab& textVocab) {
+  WordVocabIndex wvi;
+  for (auto word : tokenizeAndNormalizeText(text, localeManager_)) {
+    // Check if word exists and retrieve wordId
+    if (!textVocab.getId(word, &wvi)) {
+      continue;
+    }
+    WordIndex currentWordId = wvi.get();
+    // Emplaces or increases the docLengthMap
+    if (docLengthMap_.contains(docId)) {
+      ++docLengthMap_[docId];
+    } else {
+      docLengthMap_[docId] = 1;
+    }
+    ++totalDocumentLength_;
+    // Check if wordId already is a key in the InvertedIndex
+    if (invertedIndex_.contains(currentWordId)) {
+      // Check if the word is seen in a new context or not
+      InnerMap& innerMap = invertedIndex_.find(currentWordId)->second;
+      auto ret = innerMap.find(docId);
+      // Seen in new context
+      if (ret == innerMap.end()) {
+        innerMap.emplace(docId, 1);
+        // Seen in known context
+      } else {
+        ret->second += 1;
+      }
+      // New word never seen before
+    } else {
+      invertedIndex_.emplace(currentWordId, InnerMap{{docId, 1}});
+    }
```

Comment:
```suggestion
  ++invertedIndex_[currentWordId][docId};
```
Assuming that the `InnerMap` also is a hashMap.

Reply:
f752ad6a26f2b3595c41f43edba66a51b38f4231